include/aws/iotdevice/secure_tunneling.h

```diff
@@ -6,120 +6,260 @@
+ *
+ * Contains connection properties for the creation of a Secure Tunnel
+ */
 struct aws_secure_tunnel_options {
     struct aws_allocator *allocator;
+
+    /**
+     * Host to establish Secure Tunnel connection to
+     */
+    struct aws_byte_cursor endpoint_host;
+
+    /**
+     * Secure Tunnel bootstrap to use whenever Secure Tunnel establishes a connection
+     */
     struct aws_client_bootstrap *bootstrap;
+
+    /**
+     * Socket options to use whenever this Secure Tunnel establishes a connection
+     */
     const struct aws_socket_options *socket_options;
+
+    /**
+     * (Optional) Http proxy options to use whenever this Secure Tunnel establishes a connection
+     */
     const struct aws_http_proxy_options *http_proxy_options;
 
+    /**
+     * Access Token used to establish a Secure Tunnel connection
+     */
     struct aws_byte_cursor access_token;
-    enum aws_secure_tunneling_local_proxy_mode local_proxy_mode;
-    struct aws_byte_cursor endpoint_host;
+
+    /**
+     * (Optional) Client Token used to re-establish a Secure Tunnel connection after the one-time use access token has
+     * been used. If one is not provided, it will automatically be generated and re-used on subsequent reconnects.
+     */
+    struct aws_byte_cursor client_token;
+
     const char *root_ca;
 
+    aws_secure_tunnel_message_received_fn *on_message_received;
+
+    void *user_data;
+
+    enum aws_secure_tunneling_local_proxy_mode local_proxy_mode;
+
     aws_secure_tunneling_on_connection_complete_fn *on_connection_complete;
     aws_secure_tunneling_on_connection_shutdown_fn *on_connection_shutdown;
     aws_secure_tunneling_on_send_data_complete_fn *on_send_data_complete;
-    aws_secure_tunneling_on_data_receive_fn *on_data_receive;
     aws_secure_tunneling_on_stream_start_fn *on_stream_start;
     aws_secure_tunneling_on_stream_reset_fn *on_stream_reset;
     aws_secure_tunneling_on_session_reset_fn *on_session_reset;
     aws_secure_tunneling_on_termination_complete_fn *on_termination_complete;
-
-    void *user_data;
 };
 
-/* deprecated: "_config" is renamed "_options" for consistency with similar code in the aws-c libraries */
-#define aws_secure_tunneling_connection_config aws_secure_tunnel_options
+/**
+ * Signature of callback to invoke when a DISCONNECT is fully written to the socket (or fails to be)
+ */
+typedef void(aws_secure_tunnel_disconnect_completion_fn)(int error_code, void *complete_ctx);
 
 /**
- * Persistent storage for aws_secure_tunnel_options.
+ * Public completion callback options for the DISCONNECT operation
  */
-struct aws_secure_tunnel_options_storage;
+struct aws_secure_tunnel_disconnect_completion_options {
+    aws_secure_tunnel_disconnect_completion_fn *completion_callback;
+    void *completion_user_data;
+};
+
+/* deprecated: "_config" is renamed "_options" for consistency with similar code in the aws-c libraries */
+#define aws_secure_tunneling_connection_config aws_secure_tunnel_options
```

Comment:
I've seen Mike use this "one cool trick to change a name without breaking API" change before. It feels safe to remove since we're changing the API in other fun ways.